recipes/vasp/v6.4.2/gh200/compilers.yaml

```diff
@@ -0,0 +1,9 @@
+bootstrap:
+  spec: gcc@12.2
+gcc:
+  specs:
+  - gcc@12.2
```

Comment:
The nvhpc, gcc and cuda versions are all tied together. The nvhpc version I tested with comes with bundled cuda 12.2, which only supports up to gcc 12.2 as well. Unfortunately, spack doesn't recognize this, so I'm using a matching cuda spec to make sure MPI is built with the same cuda version.
The spack architecture flags for all gcc 12 versions are the same and only gcc 13 has neoverse v2 specific optimizations. Technically, a neoverse v2 specific flag was added in gcc 12.3.0, but the official gcc release notes only mention support with gcc 13. They are also missing from the documentation (https://gcc.gnu.org/onlinedocs/gcc-12.3.0/gcc/ARM-Options.html). That's why the same Arm architecture flags for all gcc 12 versions are used. These should cover the instruction set Grace is capable of, but don't quite have perfect tuning parameters.
In the case of this vasp recipe, all performance critical packages like vasp itself and Openblas are build with nvhpc anyway, which includes neoverse v2 optimizations.

Reply:
I will start by saying that the configuration is good for a first deployment - we can go with these software versions, and release an update with different cuda/gcc/nvhpc versions later if we need to.

We also need some documentation with recommended versions of compilers, runtimes, etc that we recommend all uenv use if possible. That is on me!

NVIDIA gave us feedback that 12.3 was the first version of GCC  that should be used with Grace (even if it isn't explicit support, it might well tested).

Have you tried nvhpc 24.1 or 24.3? They are both supported on Santis with Stackinator and Spack v0.21.

When a cuda spec, e.g. `cuda@12.3`, then that version will be installed and used, instead of the version inside the `nvhpc` SDK (at least it is in other projects that we use `nvhpc` to build the Fortran OpenACC code, e.g. ICON).

The most recent software versions that we haver tested with success (not guarenteed to work for all use cases) is:
```
nvhpc@24.1 / 24.3
gcc@13.2
cuda@12.4
```

Reply:
I think the advice from Nvidia with regards to GCC 12.3 only makes sense if you use `-mpcu=neoverse-v2`, which we don't when using spack (since it's an undocumented feature, so spack doesn't use it).

Using GCC 13 would be ideal, but it requires cuda 12.4, as you suggested. But nvhpc 24.3 only comes with cuda 12.3 (https://docs.nvidia.com/hpc-sdk/archive/24.3/hpc-sdk-release-notes/index.html). The way vasp is currently build, one has to use a bundled cuda version. And when I tested with stackinator, spack does not appear to find the bundled cuda version (at least when nvhpc is installed as a @gcc package and used as a compiler). I ended up removing the nvhpc dependency from the vasp package, since we would then end up with two nvhpc packages in the image. So an explicit cuda package version is used instead for the MPI package. I'd be curious to learn how other OpenACC codes solved this.

Going forward, I'd like to upgrade the compiler / cuda version. I could test nvhpc 24.3 and cuda 12.3 with gcc 12.2. It will take about two days to run the verification tests however.